src/components/upgrader/hpkUpgrader.ts

```diff
@@ -104,35 +111,53 @@ export default class HPKUpgrader extends EventEmitter implements IDeviceUpgrader
-        let totalProgressPoints = 1;
-        this._cameraManager.transport.on('upgrader/status', async message => {
-          const statusMessage =  Api.decode(message.payload, 'messagepack');
-          totalProgressPoints = statusMessage.total_points || totalProgressPoints;
-          if (statusMessage.operation === 'done' && statusMessage.reboot) {
-            resolve(true);
-          } else if (statusMessage.operation === 'done') {
-            resolve(false);
-          }
-          if (statusMessage.error_count > 0) {
-            return reject(statusMessage);
-          }
-          const elapsedPoints = statusMessage.elapsed_points || 0;
-          const progressPercentage = (elapsedPoints / totalProgressPoints) * 100;
-          this._logger.info(`Upgrading HPK: Status: ${Math.ceil(progressPercentage)}% step ${statusMessage.operation}\r`);
-          this.emit(CameraEvents.UPGRADE_PROGRESS, {
-            operation: statusMessage.operation,
-            progress: progressPercentage
-          });
+    const reboot = await this._cameraManager.api.withSubscribe(['upgrader/status'], () => new Promise((resolve, reject) => {
+      const statusMessageTimoutTime = 10000;
+      function startTimout() {
+        return setTimeout(() => {
+          reject(`Upgrading HPK: no status message within ${statusMessageTimoutTime}`);
+        }, statusMessageTimoutTime);
+      }
+
+      let totalProgressPoints = 1;
+      let messageTimoutIt = startTimout();
+      this._cameraManager.transport.on('upgrader/status', async message => {
+        clearTimeout(messageTimoutIt);
+        const statusMessage =  Api.decode(message.payload, 'messagepack');
+        totalProgressPoints = statusMessage.total_points || totalProgressPoints;
+        if (statusMessage.operation === 'done' && statusMessage.reboot) {
+          resolve(true);
+          return;
+        } else if (statusMessage.operation === 'done') {
+          resolve(false);
+          return;
+        }
+        if (statusMessage.error_count > 0) {
+          return reject(statusMessage);
+        }
+        const elapsedPoints = statusMessage.elapsed_points || 0;
+        const progressPercentage = (elapsedPoints / totalProgressPoints) * 100;
+        this._logger.info(`Upgrading HPK: Status: ${Math.ceil(progressPercentage)}% step ${statusMessage.operation}\r`);
+        this.emit(CameraEvents.UPGRADE_PROGRESS, {
+          operation: statusMessage.operation,
+          progress: progressPercentage
         });
-      })
-    );
 
-    if (shouldReboot) {
+        messageTimoutIt = startTimout();
+      });
+    }));
+
+    if (reboot) {
       await this._cameraManager.transport.stopEventLoop();
       await this._cameraManager.reboot();
-      await this._cameraManager.transport.close();
+      try {
+        await this._cameraManager.transport.close();
+      } catch (e) {
+        this._logger.info(`\nUpgrading HPK: failed closing device on reboot: ${e}\n`);
```

Comment:
Maybe you should at least set the `this._cameraManager = undefined` since after reboot you will get a new instance of camera manager from the attach event!